kratos/utilities/mortar_utilities.h

```diff
@@ -592,22 +600,328 @@ class MortarUtilities
+        GeometryType& ThisGeometry,
+        Variable<double>& ThisVariable,
+        const Matrix& ThisValue
+        )
+{
+    const double& r_area = ThisGeometry.Area()/ThisGeometry.PointsNumber();
+    
+    for (unsigned int i_node = 0; i_node < ThisGeometry.size(); ++i_node)
+    {
+        const double area_coeff = r_area/ThisGeometry[i_node].GetValue(NODAL_AREA);
+        ThisGeometry[i_node].FastGetSolutionStepValue(ThisVariable) += area_coeff * ThisValue(i_node, 0);
+    }
+}
+
+template<> 
+inline void MortarUtilities::AddValue<component_type, Historical>(
+        GeometryType& ThisGeometry,
+        component_type& ThisVariable,
+        const Matrix& ThisValue
+        )
+{
+    const double& r_area = ThisGeometry.Area()/ThisGeometry.PointsNumber();
+    
+    for (unsigned int i_node = 0; i_node < ThisGeometry.size(); ++i_node)
+    {
+        const double area_coeff = r_area/ThisGeometry[i_node].GetValue(NODAL_AREA);
+        ThisGeometry[i_node].FastGetSolutionStepValue(ThisVariable) += area_coeff * ThisValue(i_node, 0);
+    }
+}
+
+template<> 
+inline void MortarUtilities::AddValue<Variable<array_1d<double, 3>>, Historical>(
+        GeometryType& ThisGeometry,
+        Variable<array_1d<double, 3>>& ThisVariable,
+        const Matrix& ThisValue
+        )
+{
+    const double& r_area = ThisGeometry.Area()/ThisGeometry.PointsNumber();
+    
+    for (unsigned int i_node = 0; i_node < ThisGeometry.size(); ++i_node)
+    {
+        const double area_coeff = r_area/ThisGeometry[i_node].GetValue(NODAL_AREA);
+        auto& aux_vector = ThisGeometry[i_node].FastGetSolutionStepValue(ThisVariable);
+        aux_vector += area_coeff * row(ThisValue, i_node);
+    }
+}
+template<> 
+inline void MortarUtilities::AddValue<Variable<double>, NonHistorical>(
+        GeometryType& ThisGeometry,
+        Variable<double>& ThisVariable,
+        const Matrix& ThisValue
+        )
+{
+    const double& r_area = ThisGeometry.Area()/ThisGeometry.PointsNumber();
+    
+    for (unsigned int i_node = 0; i_node < ThisGeometry.size(); ++i_node)
+    {
+        const double area_coeff = r_area/ThisGeometry[i_node].GetValue(NODAL_AREA);
+        ThisGeometry[i_node].GetValue(ThisVariable) += area_coeff * ThisValue(i_node, 0);
+    }
+}
+
+template<> 
+inline void MortarUtilities::AddValue<component_type, NonHistorical>(
+        GeometryType& ThisGeometry,
+        component_type& ThisVariable,
+        const Matrix& ThisValue
+        )
+{
+    const double& r_area = ThisGeometry.Area()/ThisGeometry.PointsNumber();
+    
+    for (unsigned int i_node = 0; i_node < ThisGeometry.size(); ++i_node)
+    {
+        const double area_coeff = r_area/ThisGeometry[i_node].GetValue(NODAL_AREA);
+        ThisGeometry[i_node].GetValue(ThisVariable) += area_coeff * ThisValue(i_node, 0);
+    }
+}
+
+template<> 
+inline void MortarUtilities::AddValue<Variable<array_1d<double, 3>>, NonHistorical>(
```

Comment:
this weight the addition by an area. Reflect it in the name...

something like 
AddAreaWeightedValue

also note that it is not threadsafe, **and that it is only correct for triangles and tets** (not even for quads)

Reply:
Ok, in any case I am doing right now the implicit version and I will remove this methods